File: java/src/test/java/com/google/fhir/protogen/FhirPackageTest.java
//    Copyright 2020 Google LLC
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

package com.google.fhir.protogen;

import static com.google.common.truth.Truth.assertThat;
import static junit.framework.Assert.assertTrue;
import static org.junit.Assert.assertFalse;

import com.google.common.testing.EqualsTester;
import com.google.fhir.proto.PackageInfo;
import java.io.IOException;
import java.util.ArrayList;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/**
 * Tests for FhirPackage.
 */
@RunWith(JUnit4.class)
public final class FhirPackageTest {

  @Test
  public void equalityTest() throws Exception {
    new EqualsTester()
        .addEqualityGroup(
            new FhirPackage(
                PackageInfo.newBuilder().setProtoPackage("foo").build(),
                new ArrayList<>(),
                new ArrayList<>(),
                new ArrayList<>()),
            new FhirPackage(
                PackageInfo.newBuilder().setProtoPackage("foo").build(),
                new ArrayList<>(),
                new ArrayList<>(),
                new ArrayList<>()))
        .addEqualityGroup(
            new FhirPackage(
                PackageInfo.newBuilder().setProtoPackage("bar").build(),
                new ArrayList<>(),
                new ArrayList<>(),
                new ArrayList<>()),
            new FhirPackage(
                PackageInfo.newBuilder().setProtoPackage("bar").build(),
                new ArrayList<>(),
                new ArrayList<>(),
                new ArrayList<>()))
        .testEquals();
  }

  private static final int R4_DEFINITIONS_COUNT = 653;
  private static final int R4_CODESYSTEMS_COUNT = 1062;
  private static final int R4_VALUESETS_COUNT = 1316;

  @Test
  public void loadTest() throws IOException {
    FhirPackage fhirPackage = FhirPackage.load("spec/fhir_r4_package.zip");
    assertThat(fhirPackage.packageInfo.getProtoPackage()).isEqualTo("google.fhir.r4.core");
    assertThat(fhirPackage.structureDefinitions).hasSize(R4_DEFINITIONS_COUNT);
    assertThat(fhirPackage.codeSystems).hasSize(R4_CODESYSTEMS_COUNT);
    assertThat(fhirPackage.valueSets).hasSize(R4_VALUESETS_COUNT);
  }

  @Test
  public void filterTest() throws IOException {
    FhirPackage fhirPackage = FhirPackage.load("spec/fhir_r4_package.zip");
    assertTrue(
        fhirPackage.structureDefinitions.stream()
            .anyMatch(def -> def.getId().getValue().equals("Patient")));

    FhirPackage filteredPackage =
        fhirPackage.filterResources(def -> !def.getId().getValue().equals("Patient"));
    assertFalse(
        filteredPackage.structureDefinitions.stream()
            .anyMatch(def -> def.getId().getValue().equals("Patient")));
    assertThat(filteredPackage.structureDefinitions).hasSize(R4_DEFINITIONS_COUNT - 1);
  }
}

File: py/google/fhir/utils/fhir_package_test.py
#
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Test fhir_package functionality."""

import os
from absl import flags

from absl.testing import absltest
from proto.google.fhir.proto import profile_config_pb2
from google.fhir.utils import fhir_package

FLAGS = flags.FLAGS

_R4_DEFINITIONS_COUNT = 653
_R4_CODESYSTEMS_COUNT = 1062
_R4_VALUESETS_COUNT = 1316
_R4_SEARCH_PARAMETERS_COUNT = 1385


class FhirPackageTest(absltest.TestCase):

  def testFhirPackageEquality_withEqualOperands_succeeds(self):
    lhs = fhir_package.FhirPackage(
        package_info=profile_config_pb2.PackageInfo(proto_package='Foo'),
        structure_definitions=[],
        search_parameters=[],
        code_systems=[],
        value_sets=[])
    rhs = fhir_package.FhirPackage(
        package_info=profile_config_pb2.PackageInfo(proto_package='Foo'),
        structure_definitions=[],
        search_parameters=[],
        code_systems=[],
        value_sets=[])
    self.assertEqual(lhs, rhs)

  def testFhirPackageEquality_withNonEqualOperands_succeeds(self):
    lhs = fhir_package.FhirPackage(
        package_info=profile_config_pb2.PackageInfo(proto_package='Foo'),
        structure_definitions=[],
        search_parameters=[],
        code_systems=[],
        value_sets=[])
    rhs = fhir_package.FhirPackage(
        package_info=profile_config_pb2.PackageInfo(proto_package='Bar'),
        structure_definitions=[],
        search_parameters=[],
        code_systems=[],
        value_sets=[])
    self.assertNotEqual(lhs, rhs)

  def testFhirPackageLoad_withValidFhirPackage_succeeds(self):
    package_filepath = os.path.join(
        FLAGS.test_srcdir, 'com_google_fhir/spec/fhir_r4_package.zip')
    package = fhir_package.FhirPackage.load(package_filepath)
    self.assertEqual(package.package_info.proto_package, 'google.fhir.r4.core')
    self.assertLen(package.structure_definitions, _R4_DEFINITIONS_COUNT)
    self.assertLen(package.code_systems, _R4_CODESYSTEMS_COUNT)
    self.assertLen(package.value_sets, _R4_VALUESETS_COUNT)
    self.assertLen(package.search_parameters, _R4_SEARCH_PARAMETERS_COUNT)


if __name__ == '__main__':
  absltest.main()
